File: EyeAICore/src/audio/DepthAudioSourceData.cpp
#include "EyeAICore/audio/DepthAudioSourceData.hpp"
#include <cmath>
#include <numbers>
#include <vector>

std::vector<short>
createAudioData(float base_frequency, float duration, int sample_rate) {
	const float PI = std::numbers::pi;
    const int numSamples = static_cast<int>(static_cast<double>(sample_rate) * duration);
    const float amplitude = 0.8f;
    const float decay_rate = 8.0f; // Steuert wie schnell der Klick abklingt
    
    std::vector<short> samples(numSamples);
    
    for (int i = 0; i < numSamples; ++i) {
        float t = static_cast<float>(i) / static_cast<float>(sample_rate);
        
        // Exponentieller Decay für natürlichen Klick-Sound
        float envelope = std::exp(-decay_rate * t);
        
        // Mischung aus Grundfrequenz und Harmonischen für prägnanten Sound
        float wave = std::sin(2.0f * PI * base_frequency * t) +
                     0.5f * std::sin(2.0f * PI * base_frequency * 2.0f * t) +
                     0.25f * std::sin(2.0f * PI * base_frequency * 3.0f * t);
        
        samples[i] = static_cast<short>(
            amplitude * envelope * wave * 32760 / 1.75f // Normalisierung wegen der Harmonischen
        );
    }
    
    return samples;
}

File: eye-ai-core-rs/src/audio/depth_audio_source_data.rs
use crate::audio::Vec3;
use alto::Mono;

pub struct DepthAudioSourceData {
	pub frequency: f32,
	pub duration: f32,
	pub sample_rate: usize,
	pub position: Vec3,
	pub samples: Vec<Mono<i16>>,
}

impl DepthAudioSourceData {
	pub fn new(frequency: f32, duration: f32, sample_rate: usize, position: Vec3) -> Self {
		Self {
			frequency,
			duration,
			sample_rate,
			position,
			samples: create_audio_data(frequency, duration, sample_rate),
		}
	}
}

fn create_audio_data(base_frequency: f32, duration: f32, sample_rate: usize) -> Vec<Mono<i16>> {
	const TWO_PI: f32 = std::f32::consts::PI * 2.0;

	let num_samples = number_of_samples(sample_rate, duration);
	let amplitude: f32 = 0.8;
	let decay_rate: f32 = 8.0;

	let mut samples: Vec<Mono<i16>> = Vec::new();
	samples.resize(num_samples, Mono { center: 0 });
	for (i, sample) in samples.iter_mut().enumerate() {
		let t = i as f32 / sample_rate as f32;

		let envelope = (-decay_rate * t).exp();

		let wave = (TWO_PI * base_frequency * t).sin()
			+ 0.5 * (TWO_PI * base_frequency * 2.0 * t).sin()
			+ 0.25 * (TWO_PI * base_frequency * 3.0 * t).sin();

		*sample = Mono {
			center: ((amplitude * envelope * wave * 32760.0) / 1.75) as i16,
		};
	}

	samples
}

fn number_of_samples(sample_rate: usize, duration: f32) -> usize {
	(sample_rate as f32 * duration) as usize
}
